src/dbnode/storage/series/buffer.go

```diff
@@ -181,287 +167,268 @@ func (b *dbBuffer) Write(
-		if len(streams) != 1 {
-			// Should never happen as the call to merge above should result in only a single
-			// stream being present.
-			err = errMoreThanOneStreamAfterMerge
-			return
-		}
+	err = persistFn(id, tags, segment, checksum)
+	if err != nil {
+		return FlushOutcomeErr, err
+	}
 
-		// Direct indexing is safe because canRead guarantees us at least one stream
-		res = streams[0]
-	})
+	if bucket, exists := buckets.writableBucket(WarmWrite); exists {
+		bucket.version = version
+	}
 
-	return res, err
+	return FlushOutcomeFlushedToDisk, nil
 }
 
 func (b *dbBuffer) ReadEncoded(ctx context.Context, start, end time.Time) [][]xio.BlockReader {
 	// TODO(r): pool these results arrays
 	var res [][]xio.BlockReader
-	b.forEachBucketAsc(func(bucket *dbBufferBucket) {
-		if !bucket.canRead() {
-			return
-		}
-		if !start.Before(bucket.start.Add(b.blockSize)) {
-			return
+
+	for _, blockStart := range b.inOrderBlockStarts {
+		if !blockStart.Before(end) || !start.Before(blockStart.Add(b.blockSize)) {
+			continue
 		}
-		if !bucket.start.Before(end) {
-			return
+
+		bv, exists := b.bucketVersionsAt(blockStart)
+		if !exists {
+			// Invariant violated. This means the keys in the bucket map does
+			// not match the sorted keys cache, which should never happen.
+			log := b.opts.InstrumentOptions().Logger()
+			log.Errorf(errBucketMapCacheNotInSync)
+			return nil
 		}
 
-		res = append(res, bucket.streams(ctx))
+		if streams := bv.streams(ctx, WarmWrite); len(streams) > 0 {
+			res = append(res, streams)
+		}
+		if streams := bv.streams(ctx, ColdWrite); len(streams) > 0 {
+			res = append(res, streams)
+		}
 
 		// NB(r): Store the last read time, should not set this when
 		// calling FetchBlocks as a read is differentiated from
 		// a FetchBlocks call. One is initiated by an external
 		// entity and the other is used for streaming blocks between
 		// the storage nodes. This distinction is important as this
 		// data is important for use with understanding access patterns, etc.
-		bucket.setLastRead(b.nowFn())
-	})
+		bv.setLastRead(b.nowFn())
+	}
 
 	return res
 }
 
 func (b *dbBuffer) FetchBlocks(ctx context.Context, starts []time.Time) []block.FetchBlockResult {
 	var res []block.FetchBlockResult
 
-	b.forEachBucketAsc(func(bucket *dbBufferBucket) {
-		if !bucket.canRead() {
-			return
+	for _, start := range starts {
+		buckets, ok := b.bucketVersionsAt(start)
+		if !ok {
+			continue
 		}
-		found := false
-		// starts have only a few items, linear search should be okay time-wise to
-		// avoid allocating a map here.
-		for _, start := range starts {
-			if start.Equal(bucket.start) {
-				found = true
-				break
-			}
+
+		if streams := buckets.streams(ctx, WarmWrite); len(streams) > 0 {
+			res = append(res, block.NewFetchBlockResult(start, streams, nil))
 		}
-		if !found {
-			return
+
+		if streams := buckets.streams(ctx, ColdWrite); len(streams) > 0 {
+			res = append(res, block.NewFetchBlockResult(start, streams, nil))
```

Comment:
Hm don't think this is correct, we should only have a single `FetchBlockResult` for a single block `start` time (e.g. you should append the streams together then create one of these).

Maybe we should have some type of test somewhere that enforces this too (since it's obvious this isn't enforced by the tests, which it should be - an integration test would be good for this case obviously).

Reply:
Ah thanks for catching this, I feel like this would have been an annoying bug to figure out later on. I'll include an integration test in a separate PR.

Reply:
Sounds good, yeah definitely one of those that will likely break in the future unless we add a test soon.  Thanks 👍